internal/storage/rocksdb/store.go

```diff
@@ -389,37 +423,67 @@ func (rdb *rocksDB) generateSST(snap *gorocksdb.Snapshot, cf *gorocksdb.ColumnFa
 	return os.Open(fileName)
 }
 
+type checkPointSnapshot struct {
+	tar *utils.StreamingTar
+	dir string
+}
+
+func (r *checkPointSnapshot) Read(p []byte) (n int, err error) {
+	return r.tar.Read(p)
+}
+
+func (r *checkPointSnapshot) Close() error {
+	r.tar.Close()
+	return os.RemoveAll(r.dir)
+}
+
 func (rdb *rocksDB) GetSnapshot() (io.ReadCloser, error) {
 	defer rdb.opts.statsCli.Timing("rocksdb.snapshot.get.latency.ms", time.Now())
-	snap := rdb.db.NewSnapshot()
-	defer rdb.db.ReleaseSnapshot(snap)
 
 	sstDir, err := storage.CreateTempFolder(rdb.opts.sstDirectory, sstPrefix)
 	if err != nil {
 		rdb.opts.lgr.Error("GetSnapshot: Failed to create temporary dir", zap.Error(err))
 		return nil, err
 	}
-	defer os.RemoveAll(sstDir)
 
-	normalSSTFile, err := rdb.generateSST(snap, rdb.normalCF, sstDir)
+	checkpoint, err := rdb.db.NewCheckpoint()
 	if err != nil {
-		rdb.opts.lgr.Error("GetSnapshot: Failed to generate sst file", zap.Error(err))
+		rdb.opts.lgr.Error("GetSnapshot: Failed to create new checkpoint", zap.Error(err))
 		return nil, err
 	}
 
-	ttlSSTFile, err := rdb.generateSST(snap, rdb.ttlCF, sstDir)
+	checkpointDir := fmt.Sprintf("%s/checkpoint", sstDir)
+	err = checkpoint.CreateCheckpoint(checkpointDir, 0)
 	if err != nil {
-		rdb.opts.lgr.Error("GetSnapshot: Failed to generate ttl sst file", zap.Error(err))
+		rdb.opts.lgr.Error("GetSnapshot: Failed to make checkpoint", zap.Error(err))
+		return nil, err
+	}
+
+	//check that checkpoint was successfully created
+	if created, _ := exists(checkpointDir); !created {
+		err = fmt.Errorf("checkpoint.CreateCheckpoint failed")
+		rdb.opts.lgr.Error("GetSnapshot: Checkpoint dir was not created", zap.Error(err))
 		return nil, err
 	}
 
-	tarF, err := utils.CreateStreamingTar(normalSSTFile, ttlSSTFile)
+	var files []*os.File
+	filepath.WalkDir(checkpointDir, func(path string, fi fs.DirEntry, err error) error {
+		if !fi.IsDir() {
+			f, err := os.Open(path)
+			if err != nil {
+				return err
+			}
+			files = append(files, f)
+		}
+		return nil
+	})
+
+	tarF, err := utils.CreateStreamingTar(files...)
 	if err != nil {
 		rdb.opts.lgr.Error("GetSnapshot: Failed to archive sst files", zap.Error(err))
 		return nil, err
 	}
-
-	return tarF, nil
+	return &checkPointSnapshot{tar: tarF, dir: sstDir}, nil
```

Comment:
Please check if maybe we should pass `checkpointDir` for the `dir` param. Though it should not matter since `sstDir` is the parent folder anyways, but then its lot more clearer. Or else, short circuit the `sstDir` altogether.

Reply:
Well, there is a reason, why its this way. The rocksDB checkpoint function needs a tmp directory, but it shouldn't exist. In that case, we would need to call `storage.CreateTempFolder` followed by `delete` and then passing it to the checkpoint dir; IMO this was making it complicated. Taking a subfolder makes it way cleaner and handles it better. The reason to pass `sstDir` here is so that the parent tmp folder is automatically cleaned up otherwise we would end up with lot of empty folders in the sst directory.